### PR TITLE
Fix broken category search

### PR DESCRIPTION
**Pull request by NauelSerraino**
Hi @GalenReich, the new version is ready, major changes are:

- Removal of `filing_type` (`&category=..`) argument, substituted by `filing_form` (`&forms=...`)
- Additional `single_forms` argument to select one or more elements from the chosen `filing_form`

Minor changes are:

- Update of the constants file removing the verbose version of the [TEXT_SEARCH_CATEGORY_FORM_GROUPINGS](https://github.com/bellingcat/EDGAR/blob/3c2890a4863b2f734a462fbbbbcae6fd51358648/edgar_tool/constants.py#L1393) and substituting those with the ones from the old `TEXT_SEARCH_FILING_CATEGORIES_MAPPING` (in `TEXT_SEARCH_FILING_VS_MAPPING_CATEGORIES_MAPPING`, there is still a mapping to retrive the original one, if needed)
- Update of the README to specify how to handle `single_forms`
- Changed `,` with `;` in `Destination file must have one of the following extensions:` because it was a bit confusing

I hope that everything is clear, if you need further explanations feel free to ask!

### CODE REVIEW

File: edgar_tool/constants.py [hunk 1]
```diff
@@ -1390,10 +1389,9 @@
     },
 }
 
-# These groupings are not yet used, but are included for completeness
 TEXT_SEARCH_CATEGORY_FORM_GROUPINGS = {
     #    "Exclude insider equity awards, transactions, and ownership (Section 16 Reports)": ["-3","-4","-5"], # todo: work out how to exclude these
-    "All annual, quarterly, and current reports": [
+    "all_annual_quarterly_and_current_reports": [
```
Comment (GalenReich) [inline edgar_tool/constants.py:1394]: Why the change to snake case?

File: edgar_tool/text_search.py [hunk 2]
```diff
@@ -197,7 +194,7 @@ def _generate_request_args(
 
         :param keywords: Search keywords to input in the "Document word or phrase" field
         :param entity_id: Entity/Person name, ticker, or CIK number to input in the "Company name, ticker, or CIK" field
-        :param filing_type: Filing category to select from the dropdown menu, defaults to None
+        :param filing_form: Group to select within the filing category dropdown menu, defaults to None
```
Comment (GalenReich) [inline edgar_tool/text_search.py:197]: This was a pre-existing documentation fault, but I think referring to the "dropdown menu" won't be clear to a user of the tool on the command line (as there is no dropdown).

File: edgar_tool/text_search.py [hunk 3]
```diff
@@ -224,10 +221,14 @@ def _generate_request_args(
...
-                filing_type
-            ]
+        if filing_form and not single_forms:
+            request_args["forms"] = ",".join(
+                TEXT_SEARCH_CATEGORY_FORM_GROUPINGS[filing_form]
+            )
+            request_args["forms"] = request_args["forms"].replace(" ", "-")
+        if filing_form and single_forms:
+            request_args["forms"] = ",".join(single_forms)
+            request_args["forms"] = request_args["forms"].replace(" ", "-")
```
Comment (GalenReich) [inline edgar_tool/text_search.py:231]: This is another place where the behaviour could change to allow giving `single_forms` without `filing_form`. I really like the idea of combining them in the instance that both are provided!

File: edgar_tool/cli.py [hunk 4]
```diff
@@ -46,15 +47,27 @@ def _validate_text_search_args(
...
+            "single_forms can only be used when filing_form is specified"
         )
+    if single_forms:
+        invalid_forms = [
+            form for form in single_forms if form not in TEXT_SEARCH_CATEGORY_FORM_GROUPINGS[filing_form]
+        ]
+        if filing_form and invalid_forms:
+            raise ValueError(
+                f"Single forms must be one of: {TEXT_SEARCH_CATEGORY_FORM_GROUPINGS[filing_form]}"
+            )
```
[the remark this answers is not in the document]
Comment (NauelSerraino) [reply edgar_tool/cli.py:70]: There are basically 2 scenarios for this:

### First if statement (`if single_forms and not filing_form:`)

> `single_form=True` and `filing_form=False`  -> ❌ (you need first to provide the `filing_form` in order to filter a subset of it through the `single_form` list)

```bash
(edgar) PS C:\Users\\github_projects\EDGAR> poetry run edgar-tool text_search Hurricane --single_forms "['1-K', '1-SA']"                                                         
ValueError: single_forms can only be used when filing_form is specified

```
### Second if statement (`if single_forms:`)

> This `if` gets activated only if the previous statement has not raised the error (meaning that `filing_form` is provided). `invalid_forms` checks if the list provided form the user belongs to the `filing_form`; returns True if the forms provided do not belong to the subset of the selected `filing_form` (Example: user choose `filing_form = "beneficial_ownership_reports"` which admits `["SC 13D", "SC 13G", "SC14D1F"]`, user though provide `single_form = "["1-K", "1-SA"]"` which do not belong to `beneficial_ownership_reports` so error is raised ❌)

```bash
(edgar) PS C:\Users\cg14373\github_projects\EDGAR> poetry run edgar-tool text_search Hurricane --filing_form "beneficial_ownership_reports" --single_forms "['1-K', '1-SA']"            

ValueError: Single forms must be one of: ['SC 13D', 'SC 13G', 'SC14D1F']
```
---
Of course there is a third scenario where both arguments are evaluated and the list provided by the user belongs to the available options ✅

```bash
(edgar) PS C:\Users\cg14373\github_projects\EDGAR> poetry run edgar-tool text_search Hurricane --filing_form "all_annual_quarterly_and_current_reports" --single_forms "['1-K', '1-SA']"    


Requesting URL: https://efts.sec.gov/LATEST/search-index?q=Hurricane&dateRange=custom&startdt=2019-05-22&enddt=2024-05-20&page=1&forms=1-K%2C1-SA
Waiting 0.1396511268655101 seconds for the request to complete...
Successfully fetched URL: https://efts.sec.gov/LATEST/search-index?q=Hurricane&dateRange=custom&startdt=2019-05-22&enddt=2024-05-20&page=1&forms=1-K%2C1-SA
Less than 10000 (27) results found for range 2019-05-22 00:00:00 -> 2024-05-20 00:00:00, returning search request string...
Requesting URL: https://efts.sec.gov/LATEST/search-index?q=Hurricane&dateRange=custom&startdt=2019-05-22&enddt=2024-05-20&page=1&forms=1-K%2C1-SA
Waiting 0.13544497374766112 seconds for the request to complete...
Successfully fetched URL: https://efts.sec.gov/LATEST/search-index?q=Hurricane&dateRange=custom&startdt=2019-05-22&enddt=2024-05-20&page=1&forms=1-K%2C1-SA
Found 27 / 100 = 1 pages
Requesting URL: https://efts.sec.gov/LATEST/search-index?q=Hurricane&dateRange=custom&startdt=2019-05-22&enddt=2024-05-20&page=1&forms=1-K%2C1-SA&page=1
Waiting 0.1469443806017155 seconds for the request to complete...
Successfully fetched URL: https://efts.sec.gov/LATEST/search-index?q=Hurricane&dateRange=custom&startdt=2019-05-22&enddt=2024-05-20&page=1&forms=1-K%2C1-SA&page=1
Successfully wrote data to edgar_search_results_20240520_153725.csv.
```

File: edgar_tool/constants.py (hunk 1 above)
Comment (NauelSerraino) [reply edgar_tool/constants.py:1394]: I opted for snake case because it's easier for users to remember. Since it's case sensitive, even a minor error like a misplaced comma would cause an error.

A more sophisticated solution could be implemented using a basic NLP algorithm or regex, but I preferred to keep it simple. 

If you prefer I can revert it with the original keys.

File: edgar_tool/text_search.py (hunk 2 above)
Comment (NauelSerraino) [reply edgar_tool/text_search.py:197]: I can change it to:
`:param filing_form: Group to select within the filing category, defaults to None`

File: edgar_tool/text_search.py (hunk 3 above)
Comment (NauelSerraino) [reply edgar_tool/text_search.py:231]: I don't get this. You're saying you prefer to have the possibility to run `text_search` with the `single_forms` alone, is that correct?

File: edgar_tool/cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (GalenReich) [reply edgar_tool/cli.py:70]: That's good to clarify. Though my question wasn't very clear - apologies!

I'm suggesting that we allow users to specify `single_forms` only:

i.e. `poetry run edgar-tool text_search Hurricane --single_forms "['1-K', '1-SA', 'SC14D1F']"`

The request going to EDGAR converts the groups into lists of forms:

https://efts.sec.gov/LATEST/search-index?q=Hurricane&[...]&forms=1-K%2C1-SA%2CSC14D1F&page=1

I don't see why we would limit the range of forms to be constrained by one of the categories.

The categories act as a shorthand that makes it easier to get common groupings of form types, but someone may want to grab some arbitrary selection of form types that don't fall into one category, which would not be possible with the current approach.

File: edgar_tool/constants.py (hunk 1 above)
Comment (GalenReich) [reply edgar_tool/constants.py:1394]: This is a good reason! You are right that we should definitely prioritize user experience

My only hesitation is if EDGAR updates the categories in the future. Updating the tool would take a bit more work. Currently the form groupings are just copied from some javascript objects. Perhaps the best solution is to revert for `TEXT_SEARCH_CATEGORY_FORM_GROUPINGS` but allow users to specify the snake_case options as well using the new mapping object you created `TEXT_SEARCH_FILING_VS_MAPPING_CATEGORIES_MAPPING`.

Then for the notebook application, we can use the more readable options for a drop-down. But it means command line users wont be disadvantaged. What do you think?

File: edgar_tool/text_search.py (hunk 2 above)
Comment (GalenReich) [reply edgar_tool/text_search.py:197]: Oh wait, I'm being silly. This relates to the SEC's EDGAR front-end, not our tool, it's replicating the inputs. No need to do anything here!

File: edgar_tool/text_search.py (hunk 3 above)
Comment (GalenReich) [reply edgar_tool/text_search.py:231]: Yes, exactly. I've put a bit more detail on the other comment: https://github.com/bellingcat/EDGAR/pull/25#discussion_r1606816657

File: edgar_tool/cli.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (NauelSerraino) [reply edgar_tool/cli.py:70]: Ok now it' clear - thanks!

For sure that is something that can be done, I'll work on it.

File: edgar_tool/constants.py (hunk 1 above)
Comment (NauelSerraino) [reply edgar_tool/constants.py:1394]: Sounds good to me.

What if I revert `TEXT_SEARCH_CATEGORY_FORM_GROUPINGS` and use `TEXT_SEARCH_FILING_VS_MAPPING_CATEGORIES_MAPPING` to map the snake case input from the user into the "verbose" category?

Comment (GalenReich) [reply edgar_tool/constants.py:1394]: Perfect - thank you!

File: edgar_tool/constants.py [hunk 5]
```diff
@@ -1391,7 +1391,7 @@
 
 TEXT_SEARCH_CATEGORY_FORM_GROUPINGS = {
     #    "Exclude insider equity awards, transactions, and ownership (Section 16 Reports)": ["-3","-4","-5"], # todo: work out how to exclude these
-    "all_annual_quarterly_and_current_reports": [
+    "All annual, quarterly, and current reports": [
```
Comment (NauelSerraino) [inline edgar_tool/constants.py:1394]: Reconverted from snake format to the normal one

File: edgar_tool/cli.py [hunk 6]
```diff
@@ -132,7 +129,7 @@ def text_search(
             scraper.text_search(
                 keywords=keywords,
                 entity_id=entity_id,
-                filing_form=filing_form,
+                filing_form=TEXT_SEARCH_FILING_VS_MAPPING_CATEGORIES_MAPPING.get(filing_form),
```
Comment (NauelSerraino) [inline edgar_tool/cli.py:132]: here from snake case it gets converted into the normal case

File: edgar_tool/constants.py [hunk 7]
```diff
@@ -1604,20 +1604,20 @@
...
+    "all_section_16": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[1],
+    "beneficial_ownership_reports": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[2],
+    "exempt_offerings": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[3],
+    "registration_statements": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[4],
+    "filing_review_correspondence": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[5],
+    "sec_orders_and_notices": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[6],
+    "proxy_materials": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[7],
+    "tender_offers_and_going_private_tx": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[8],
+    "trust_indentures": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[9],
 }
```
Comment (NauelSerraino) [inline edgar_tool/constants.py:1623]: improved, it was previously hard-coded

File: edgar_tool/text_search.py [hunk 8]
```diff
@@ -221,15 +222,26 @@ def _generate_request_args(
...
             request_args["forms"] = ",".join(single_forms)
             request_args["forms"] = request_args["forms"].replace(" ", "-")
 
+        # Case where both filing_form and single_forms are specified
+        if filing_form and single_forms:
+            all_forms = TEXT_SEARCH_CATEGORY_FORM_GROUPINGS[filing_form] + single_forms
+            all_forms = list(dict.fromkeys(all_forms))  # Remove potential duplicates
+            request_args["forms"] = ",".join(all_forms)
+            request_args["forms"] = request_args["forms"].replace(" ", "-")
+
```
Comment (NauelSerraino) [inline edgar_tool/text_search.py:244]: 'core' of the feature, which handles all the scenarios

File: edgar_tool/text_search.py [hunk 9]
```diff
@@ -315,6 +327,8 @@ def _generate_search_requests(
 
         :param keywords: Search keywords to input in the "Document word or phrase" field
         :param entity_id: Entity/Person name, ticker, or CIK number to input in the "Company name, ticker, or CIK" field
+        :param filing_form: Group to select within the filing category, defaults to None
+        :param single_forms: List of single forms to search for (e.g. ['10-K', '10-Q']), defaults to None
```
Comment (NauelSerraino) [inline edgar_tool/text_search.py:331]: missing param

File: edgar_tool/text_search.py [hunk 10]
```diff
@@ -194,7 +194,8 @@ def _generate_request_args(
 
         :param keywords: Search keywords to input in the "Document word or phrase" field
         :param entity_id: Entity/Person name, ticker, or CIK number to input in the "Company name, ticker, or CIK" field
-        :param filing_form: Group to select within the filing category dropdown menu, defaults to None
+        :param filing_form: Group to select within the filing category, defaults to None
+        :param single_forms: List of single forms to search for (e.g. ['10-K', '10-Q']), defaults to None
```
Comment (NauelSerraino) [inline edgar_tool/text_search.py:198]: rewritten as accorded in the review

File: README.md [hunk 11]
```diff
@@ -46,8 +46,8 @@ edgar-tool text_search \"John Doe\" Pharmaceuticals Chemicals
 # Usage with date range and export to custom CSV file
 edgar-tool text_search Tsunami Hazards --start_date "2021-01-01" --end_date "2021-12-31" --output "results.csv"
 
-# Usage with a partial set of filing forms
-edgar-tool text_search Hurricane Damage --filing_form "all_annual_quarterly_and_current_reports" --single_forms "['1-K', '1-SA']"
+# Usage with a partial set of filing forms + single forms 
+edgar-tool text_search Hurricane Damage --filing_form "registration_statements" --single_forms "['1-K', '1-SA']"
```
Comment (NauelSerraino) [inline README.md:50]: adaptation of the README for the new feature

**Review by NauelSerraino — COMMENTED** [review]
Here I've added the case of handling both arguments together or alone. 

Hope everything is clear now :)

File: README.md [hunk 12]
```diff
@@ -46,6 +46,9 @@ edgar-tool text_search \"John Doe\" Pharmaceuticals Chemicals
 # Usage with date range and export to custom CSV file
 edgar-tool text_search Tsunami Hazards --start_date "2021-01-01" --end_date "2021-12-31" --output "results.csv"
 
+# Usage with a partial set of filing forms + single forms 
+edgar-tool text_search Hurricane Damage --filing_form "registration_statements" --single_forms "['1-K', '1-SA']"
```
Comment (NauelSerraino) [inline README.md:50]: adaptation of the README for the new feature

File: edgar_tool/cli.py [hunk 13]
```diff
@@ -46,15 +48,23 @@ def _validate_text_search_args(
...
+            f"Filing form group must be one of: {'; '.join(TEXT_SEARCH_FILING_VS_MAPPING_CATEGORIES_MAPPING.keys())}"
+    )
+    if single_forms:
+        single_list = [item for sublist in TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.values() for item in
+                       sublist]
+        invalid_forms = [form for form in single_forms if form not in single_list]
+        if invalid_forms:
+            raise ValueError(
+                f"Single forms must be one or more of: {single_list}"
+            )
```
Comment (NauelSerraino) [inline edgar_tool/cli.py:67]: here it does not raise error as before: it handles cases where only `single_forms` are present, those forms just needs to be one of all of the possible values of `TEXT_SEARCH_CATEGORY_FORM_GROUPINGS`

File: edgar_tool/constants.py [hunk 14]
```diff
@@ -1608,3 +1606,18 @@
...
+    "all_section_16": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[1],
+    "beneficial_ownership_reports": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[2],
+    "exempt_offerings": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[3],
+    "registration_statements": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[4],
+    "filing_review_correspondence": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[5],
+    "sec_orders_and_notices": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[6],
+    "proxy_materials": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[7],
+    "tender_offers_and_going_private_tx": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[8],
+    "trust_indentures": list(TEXT_SEARCH_CATEGORY_FORM_GROUPINGS.keys())[9],
+}
```
Comment (NauelSerraino) [inline edgar_tool/constants.py:1623]: improved, it was previously hard-coded

File: edgar_tool/text_search.py [hunk 15]
```diff
@@ -224,10 +222,25 @@ def _generate_request_args(
...
+        if single_forms and not filing_form:
+            request_args["forms"] = ",".join(single_forms)
+            request_args["forms"] = request_args["forms"].replace(" ", "-")
+
+        # Case where both filing_form and single_forms are specified
+        if filing_form and single_forms:
+            all_forms = TEXT_SEARCH_CATEGORY_FORM_GROUPINGS[filing_form] + single_forms
+            all_forms = list(dict.fromkeys(all_forms))  # Remove potential duplicates
+            request_args["forms"] = ",".join(all_forms)
+            request_args["forms"] = request_args["forms"].replace(" ", "-")
```
Comment (NauelSerraino) [inline edgar_tool/text_search.py:243]: 'core' of the feature, which handles all the scenarios

File: edgar_tool/constants.py (hunk 14 above)
Comment (GalenReich) [reply edgar_tool/constants.py:1623]: I actually think it is better to be hard-coded as it is a mapping.

For example, this approach could cause problems if the order of items in `TEXT_SEARCH_CATEGORY_FORM_GROUPINGS` changed for some reason.

If you could revert, that would be great.

**Review by GalenReich — CHANGES_REQUESTED** [review]

**Review by GalenReich — APPROVED** [review]
